### PR TITLE
Issue 3493: Metric for counting ZK session expirations in the Controller

### LINKED ISSUE #3493: Metric to count the ZK session expiration events
**Problem description**
ZK session expirations are a relevant event to keep track in a Pravega cluster. If not properly handled, ZK session expirations may become a source of problems. For this reason, it may be interesting for an administrator to have a straightforward way of looking into this kind of events. If multiple correlated expirations occur, it may indicate a problem in the ZK cluster.

**Problem location**
Controller.

**Suggestions for an improvement**
Add a metric to account for the number of ZK session expirations.

### PR DESCRIPTION
**Pull request by RaulGracia**
**Change log description**  
Created new metric to keep track of Zookeeper session expirations in the Controller.

**Purpose of the change**  
Fixes #3493.

**What the code does**  
A Zookeeper session expiration may be problematic if not properly handled, and in some cases, could be an indication of an issue in the ZK cluster. To be able of rapidly detecting this kind of situations in a Pravega deployment via a dashboard or alerts, we added a new metric called `controller.zookeeper.session_expiration` to account for the number of session expirations experienced by Controller instances. In turn, a session expiration in the Controller indicates that it is going to perform a graceful restart.

**How to verify it**  
Added a new test to verify that the metric works. All test should be passing as before.

Signed-off-by: Raúl Gracia <raul.gracia@emc.com>

### CODE REVIEW

**Comment by RaulGracia** [comment]
[Metric to count the ZK session expiration events](https://trello.com/c/VY9vpNg9/155-metric-to-count-the-zk-session-expiration-events)

File: controller/src/main/java/io/pravega/controller/metrics/ZookeeperConnectionMetrics.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+ */
+package io.pravega.controller.metrics;
+
+import static io.pravega.shared.MetricsNames.CONTROLLER_ZK_SESSION_EXPIRATION;
+import static io.pravega.shared.MetricsNames.globalMetricName;
+
+/**
+ * Class to encapsulate the logic to report Controller metrics related to Zookeeper.
+ */
+public class ZookeeperConnectionMetrics extends AbstractControllerMetrics {
```
Comment (fpj) [inline controller/src/main/java/io/pravega/controller/metrics/ZookeeperConnectionMetrics.java]: A minor question about the name: why connection metrics? If it is related to ZooKeeper, then it should be ZooKeeper metrics.

File: controller/src/main/java/io/pravega/controller/metrics/ZookeeperConnectionMetrics.java [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+    /**
+     * This method reports a new session expiration event in a Controller instance.
+     */
+    public void reportZKSessionExpiration() {
+        DYNAMIC_LOGGER.incCounterValue(globalMetricName(CONTROLLER_ZK_SESSION_EXPIRATION), 1);
```
Comment (fpj) [inline controller/src/main/java/io/pravega/controller/metrics/ZookeeperConnectionMetrics.java]: Does it match the dimensional way of doing metrics that we are proposing in a concurrent PR?

File: test/integration/src/test/java/io/pravega/test/integration/ControllerMetricsTest.java [hunk 3]
```diff
@@ -207,6 +210,23 @@ public void streamMetricsTest() {
         checkStatsRegisteredValues(iterations * iterations, UPDATE_STREAM_LATENCY, TRUNCATE_STREAM_LATENCY);
     }
 
+    /**
+     * This test verifies that the Controller increments the metric for Zookeeper session expiration events correctly.
+     *
+     * @throws Exception
+     */
+    @Test(timeout = 60000)
```
Comment (fpj) [inline test/integration/src/test/java/io/pravega/test/integration/ControllerMetricsTest.java]: ! minute is a long wait. Can we have a shortet timeout?

File: test/integration/src/test/java/io/pravega/test/integration/ControllerMetricsTest.java [hunk 4]
```diff
@@ -207,6 +210,23 @@ public void streamMetricsTest() {
...
+    public void zookeeperMetricsTest() throws Exception {
+        Counter zkSessionExpirationCounter = MetricRegistryUtils.getCounter(getCounterMetricName(globalMetricName(CONTROLLER_ZK_SESSION_EXPIRATION)));
+        Assert.assertNull(zkSessionExpirationCounter);
+        controllerWrapper.forceClientSessionExpiry();
+        while (zkSessionExpirationCounter == null) {
+            Thread.sleep(1000);
```
Comment (fpj) [inline test/integration/src/test/java/io/pravega/test/integration/ControllerMetricsTest.java]: Let's make this shorter to speed up the test run.

File: controller/src/main/java/io/pravega/controller/metrics/ZookeeperConnectionMetrics.java (hunk 1 above)
Comment (RaulGracia) [reply controller/src/main/java/io/pravega/controller/metrics/ZookeeperConnectionMetrics.java]: Done.

File: controller/src/main/java/io/pravega/controller/metrics/ZookeeperConnectionMetrics.java (hunk 2 above)
Comment (RaulGracia) [reply controller/src/main/java/io/pravega/controller/metrics/ZookeeperConnectionMetrics.java]: Well, this is just a counter and we just export it in the same way as all the other counters in the Controller (e.g., Streams, Transactions). When the dimensions (i.e., tags) come into play, then we may need to add them to all the Controller metrics, right?

File: test/integration/src/test/java/io/pravega/test/integration/ControllerMetricsTest.java (hunk 3 above)
Comment (RaulGracia) [reply test/integration/src/test/java/io/pravega/test/integration/ControllerMetricsTest.java]: Done.

File: test/integration/src/test/java/io/pravega/test/integration/ControllerMetricsTest.java (hunk 4 above)
Comment (RaulGracia) [reply test/integration/src/test/java/io/pravega/test/integration/ControllerMetricsTest.java]: Done.

File: controller/src/main/java/io/pravega/controller/metrics/ZookeeperConnectionMetrics.java (hunk 2 above)
Comment (kevinhan88) [reply controller/src/main/java/io/pravega/controller/metrics/ZookeeperConnectionMetrics.java]: This is the global metric so there is no additional dimensions applied (besides common tags such as host).

File: controller/src/main/java/io/pravega/controller/metrics/ZookeeperMetrics.java [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.controller.metrics;
+
+import static io.pravega.shared.MetricsNames.CONTROLLER_ZK_SESSION_EXPIRATION;
+
+/**
+ * Class to encapsulate the logic to report Controller metrics related to Zookeeper.
+ */
+public class ZookeeperMetrics extends AbstractControllerMetrics {
```
Comment (kevinhan88) [inline controller/src/main/java/io/pravega/controller/metrics/ZookeeperMetrics.java]: Make this class final.

Comment (RaulGracia) [reply controller/src/main/java/io/pravega/controller/metrics/ZookeeperMetrics.java]: Done.

**Comment by kevinhan88** [comment]
LGTM. @RaulGracia could you please hold this PR until after https://github.com/pravega/pravega/pull/3504 is merged? It will cause conflicts again and it is easier to solve conflict for this smaller PR. I try to get 3504 in before tomorrow.

**Pull request merged**